### PR TITLE
Hash attributes and Custom Elements

### PR DESCRIPTION
**Pull request by rahoulb**
This PR does the following:

- lets you define attributes with hash values - `div data: { my_attribute: "value" }` => `<div data-my-attribute="value"></div>`
- lets you create custom elements and web-components - `my_custom_element prop: "value"` => `<my-custom-element prop="value"></my-custom-element>`
- splits tagsets into separate files
- starts the process of moving functionality out of the builder into the relevant tagset (although there is still more to do)
- formats the code using StandardRB

To allow custom elements, I've had to reinstate `method_missing` on the Builder (with an associated `respond_to_missing?` method to prevent StandardRB complaining).

### CODE REVIEW

**Comment by rahoulb** [comment]
Incidentally, I've been testing against Ruby 2.7.3.  I've also switched the hashes to the 1.9 format.

**Comment by smtlaissezfaire** [comment]
Nice!  Haven't reviewed yet, will review tomorrow!

What's the thinking / use case behind the custom elements?  Doesn't Builder do this?

Also, (haven't checked your patch yet, so ignore if this is the case, but -) could we do the method missing stuff only on the custom builders / tag sets?

**Comment by rahoulb** [comment]
> What's the thinking / use case behind the custom elements? Doesn't Builder do this?

I wanted the custom elements to only be available on HTML5 documents, but the list of legal tags is filtered by the Tagset - so I had to extend it.  

> Also, (haven't checked your patch yet, so ignore if this is the case, but -) could we do the method missing stuff only on the custom builders / tag sets?

Yes - my next bit of work is to continue splitting the functionality out of the builder class and into the relevant tagset.  Currently the builder asks the tagset if the tag-name is valid and if so, the builder then asks the tagset to adjust the tag-name and attributes - before the builder finally creates the element.  I think, instead of the builder asking questions of the tagset, the tagset itself should do all the work - which means method_missing will only need to be on the HTML5 tagset.  

However, this PR is functional and I'm using my fork in some of my own work.  But I'm not sure when I'll get time to do the refactor - at least not for a few weeks anyway.  So I thought I'd publish early so you can take a look.  

Finally - thanks to Karlos who maintains Camping for reminding me about Markaby just as I was looking for something to help me author my own web-components in Ruby.

**Review by smtlaissezfaire — APPROVED** [review]
Now that's some chunky bacon!

Thanks!  Merging!

File: lib/markaby/builder.rb [hunk 1]
```diff
@@ -282,11 +255,15 @@ def initialize(*args)
       undef_method method if method_defined?(method)
     end
 
-  private
+    private
 
-    def method_missing(*args, &block)
+    def method_missing(...)
```
Comment (dentarg) [inline lib/markaby/builder.rb:260]: FYI this syntax require Ruby >= 2.7 but the markaby gem doesn't specify [`required_ruby_version`](https://guides.rubygems.org/specification-reference/#required_ruby_version) in the gemspec ([noticed](https://github.com/sinatra/sinatra/actions/runs/7146051224/job/19462956419#step:5:23) in Sinatra CI but we've applied [workaround](https://github.com/sinatra/sinatra/pull/1965))

Comment (smtlaissezfaire) [reply lib/markaby/builder.rb:260]: What's the earliest versions that sinatra requires?  Maybe this should be fixed?

Comment (smtlaissezfaire) [reply lib/markaby/builder.rb:260]: FYI @dentarg just added required_ruby_version >= 2.7 to the gemspec and release markaby 0.9.3 https://rubygems.org/gems/markaby/versions/0.9.3

Comment (dentarg) [reply lib/markaby/builder.rb:260]: Sinatra still supports Ruby 2.6. Nice, thanks for the very quick turnaround 👍

**Pull request merged**